### PR TITLE
Add websocket client

### PR DESCRIPTION
**Pull request by menarulalam**

### CODE REVIEW

File: Pipfile [hunk 1]
```diff
@@ -27,11 +27,13 @@ sphinx = "*"
 twine = "*"
 setuptools = "*"
 bump2version = "*"
+websockets = "*"
```
Comment (michaelhly) [inline Pipfile:30]: Remove from dev packages.

File: solana/rpc/providers/websocket.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
+import asyncio
```
Comment (michaelhly) [inline solana/rpc/providers/websocket.py:1]: Cite original source please.

File: solana/rpc/providers/websocket.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    async def coro_make_request(self, request_data: bytes) -> RPCResponse:
+        async with self.conn as conn:
+            await asyncio.wait_for(conn.send(request_data), timeout=self.websocket_timeout)
+            return json.loads(await asyncio.wait_for(conn.recv(), timeout=self.websocket_timeout))
+
+    def make_request(self, method: RPCMethod, *params: Any) -> RPCResponse:
+        """Make a request ot the rpc endpoint."""
+        request_id = next(self._request_counter) + 1
+        self.logger.debug(
+            "Making HTTP request. URI: %s, RequestID: %d, Method: %s, Params: %s",
```
Comment (michaelhly) [inline solana/rpc/providers/websocket.py:105]: Rename log string.

File: solana/rpc/providers/websocket.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+
+    async def coro_make_request(self, request_data: bytes) -> RPCResponse:
+        """Makes coroutine requests."""
+        async with self.conn as conn:
+            await asyncio.wait_for(conn.send(request_data), timeout=self.websocket_timeout)
+            return json.loads(await asyncio.wait_for(conn.recv(), timeout=self.websocket_timeout))
+
+    def make_request(self, method: RPCMethod, *params: Any) -> RPCResponse:
+        """Make a request ot the rpc endpoint."""
+        self.logger.debug("Making request WebSocket. URI: %s, " "Method: %s", self.endpoint_uri, method)
```
Comment (michaelhly) [inline solana/rpc/providers/websocket.py:114]: Log request_id.

**Comment by menarulalam** [comment]
Should be all done now, lmk if there is anything else you want me to change.

File: notebooks/Websockets tests.ipynb [hunk 5]
```diff
@@ -0,0 +1,313 @@
+{
```
Comment (michaelhly) [inline notebooks/Websockets tests.ipynb:1]: Delete.

File: solana/rpc/wsc.py [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+        >>> solana_client = WebSocketClient("ws://localhost:8900")
+        >>> solana_client.vote_unsubscribe(solana_client.vote_subscribe()['result']) #doctest: +SKIP
+        {'jsonrpc': '2.0', 'result': 11, 'id': 17}
+        """
+        return self._provider.make_request(RPCMethod("voteUnsubscribe"), subscription_id)
+
+    def is_connected(self) -> bool:
+        """Health check.
+
+        >>> solana_client = Client("http://localhost:8900")
```
Comment (michaelhly) [inline solana/rpc/wsc.py:162]: Wrong endpoint.

File: solana/rpc/wsc.py [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+from .providers.ws import Provider
+from .types import RPCMethod, RPCResponse
+
+
+class WebSocketClient:
+    """Websocket client class."""
+
+    def __init__(self, endpoint: str):
+        """Init API client for websocket."""
+        self._provider = Provider(endpoint)
```
Comment (michaelhly) [inline solana/rpc/wsc.py:17]: I think you need to expose the socket for this client.

File: solana/rpc/wsc.py [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+        """
+        return self._provider.make_request(RPCMethod("accountSubscribe"), str(pubkey))
+
+    def account_unsubscribe(self, account_id: int) -> RPCResponse:
+        """Unsubscribe from account change notifications.
+
+        param account_id: id of account Subscription to cancel
+
+        >>> solana_client = WebSocketClient("ws://localhost:8900")
+        >>> out = solana_client.account_subscribe(PublicKey(1))
```
Comment (michaelhly) [inline solana/rpc/wsc.py:37]: What is `out`?

File: solana/rpc/wsc.py [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+        ... '2EBVM6cB8vAAD93Ktr6Vd8p67XPbQzCJX47MpReuiCXJAtcjaxpvWpcg9Ege1Nr5Tk3a2GFrByT7WPBjdsTycY9b')
+        >>> id = out['result']
+        >>> solana_client.signature_unsubscribe(id) #doctest: +SKIP
+        {'jsonrpc': '2.0', 'result': True, 'id': 9}
+        """
+        return self._provider.make_request(RPCMethod("signatureUnsubscribe"), subscription_id)
+        # def get_account_info(self, pubkey: Union[PublicKey, str]) -> RPCResponse:
+        # return self._provider.make_request(RPCMethod("getAccountInfo"), str(pubkey))
+        # def get_account_info(self, pubkey: Union[PublicKey, str]) -> RPCResponse:
+        # return self._provider.make_request(RPCMethod("getAccountInfo"), str(pubkey))
```
Comment (michaelhly) [inline solana/rpc/wsc.py:95]: remove?

File: tests/integration_tests/test_ws_client.py [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+import pytest
+
+from solana.publickey import PublicKey
+from solana.rpc.websocket import WebSocketClient
+
+from .utils import assert_valid_response
+
+RETEST_LIMIT = 10
+
+ID = {}
```
Comment (michaelhly) [inline tests/integration_tests/test_ws_client.py:11]: What do we need a global `ID` dictionary here?

Comment (menarulalam) [reply tests/integration_tests/test_ws_client.py:11]: To unsub with old id from some things if you previously subbed

Comment (menarulalam) [reply tests/integration_tests/test_ws_client.py:11]: Should I make it into a class?

File: solana/rpc/providers/ws.py [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+
+    def is_connected(self) -> bool:
+        """Health check."""
+        try:
+            if Provider._loop is not None:
+                future = asyncio.run_coroutine_threadsafe(self.coro_make_request(bytes()), Provider._loop)
+                future.result(timeout=self.websocket_timeout)
+                return True
+        except asyncio.TimeoutError:
+            self.logger.debug("Websocket connection timed out.")
```
Comment (michaelhly) [inline solana/rpc/providers/ws.py:130]: I like how you logged the exception here, can you log exceptions everywhere else too?

File: solana/rpc/providers/ws.py [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+
+    def __str__(self) -> str:
+        """Return a string representation of the websocket connection."""
+        return "WS connection {0}".format(self.endpoint_uri)
+
+    async def coro_make_request(self, request_data: bytes) -> RPCResponse:
+        """Makes coroutine requests."""
+        async with self.conn as conn:
+            await asyncio.wait_for(conn.send(request_data), timeout=self.websocket_timeout)
+            return json.loads(await asyncio.wait_for(conn.recv(), timeout=self.websocket_timeout))
```
Comment (michaelhly) [inline solana/rpc/providers/ws.py:111]: Log possible exceptions.

File: solana/rpc/providers/ws.py [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+def _get_threaded_loop() -> asyncio.AbstractEventLoop:
+    """Get threaded loop."""
+    new_loop = asyncio.new_event_loop()
+    thread_loop = Thread(target=_start_event_loop, args=(new_loop,), daemon=True)
+    thread_loop.start()
+    return new_loop
+
+
+class PersistentWebSocket:
+    """Sets up a persistent websocket."""
```
Comment (michaelhly) [inline solana/rpc/providers/ws.py:43]: Add logger and log possible exceptions.

File: tests/integration_tests/test_ws_client.py [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+    assert_valid_response(resp)
+    if "vote" in ID:
+        ID.pop("vote")
+
+
+@pytest.mark.integration
+def test_is_connected(test_ws_client):  # pylint: disable=redefined-outer-name
+    """Test connection."""
+    resp = test_ws_client.is_connected()
+    assert resp
```
Comment (michaelhly) [inline tests/integration_tests/test_ws_client.py:140]:
```suggestion
    assert test_ws_client.is_connected()
```

File: solana/rpc/websocket.py [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+from .types import RPCMethod, RPCResponse
+
+
+class WebSocketClient:
+    """Websocket client class."""
+
+    def __init__(self, endpoint: str):
+        """Init API client for websocket."""
+        self._provider = Provider(endpoint)
+
```
Comment (michaelhly) [inline solana/rpc/websocket.py:18]: I think we also need to expose the WebSocket connection here b/c right now this client only allows you to subscribe/unsubscribe. You'll need to double-check this but I think we need something like:

```python
    def __init__(self, endpoint: str):
        """Init API client for websocket."""
        self._provider = Provider(endpoint)
        self._conn = _provider.conn

   def connection(self) -> PersistentWebSocket:
        return self._conn
```

File: solana/rpc/providers/ws.py [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+            websocket_kwargs = {}
+        else:
+            found_restricted_keys = set(websocket_kwargs.keys()).intersection(RESTRICTED_WEBSOCKET_KWARGS)
+            if found_restricted_keys:
+                raise Exception(
+                    "{0} are not allowed in websocket_kwargs, "
+                    "found: {1}".format(RESTRICTED_WEBSOCKET_KWARGS, found_restricted_keys)
+                )
+        self.conn = PersistentWebSocket(self.endpoint_uri, Provider._loop, websocket_kwargs)
+        super().__init__()
```
Comment (michaelhly) [inline solana/rpc/providers/ws.py:101]: I don't think we need this line.

Comment (michaelhly) [inline solana/rpc/providers/ws.py:101]:
```suggestion
```

Comment (michaelhly) [reply solana/rpc/providers/ws.py:101]: I don't think we need this.

File: tests/integration_tests/test_ws_client.py (hunk 14 above)
Comment (michaelhly) [reply tests/integration_tests/test_ws_client.py:140]: Actually we don't need this test, this is tested by `test_ws_client`.

File: tests/integration_tests/test_ws_client.py [hunk 17]
```diff
@@ -0,0 +1,140 @@
...
+    """Test vote unsubscribe."""
+    sub_id = ID.get("vote", test_ws_client.vote_subscribe()["result"])
+    resp = test_ws_client.vote_unsubscribe(sub_id)
+    assert_valid_response(resp)
+    if "vote" in ID:
+        ID.pop("vote")
+
+
+@pytest.mark.integration
+def test_is_connected(test_ws_client):  # pylint: disable=redefined-outer-name
```
Comment (michaelhly) [inline tests/integration_tests/test_ws_client.py:137]: We don't need this test. This is already tested by `test_ws_client`.

File: tests/integration_tests/test_ws_client.py (hunk 10 above)
Comment (menarulalam) [reply tests/integration_tests/test_ws_client.py:11]: I think it will be better to run those specific tests with a new websocket client

Comment (michaelhly) [reply tests/integration_tests/test_ws_client.py:11]: Yeah. I think running with a new websocket client is a good idea.

File: solana/rpc/websocket.py (hunk 15 above)
Comment (michaelhly) [reply solana/rpc/websocket.py:18]: https://websockets.readthedocs.io/en/stable/api.html#websockets.client.connect

**Comment by michaelhly** [comment]
Can you also add websockets as one of the required dependencies here:
https://github.com/michaelhly/solana-py/blob/0251648328159533eda1bd92a730f13603315936/setup.py#L36

File: solana/rpc/providers/ws.py [hunk 18]
```diff
@@ -0,0 +1,131 @@
...
+            self.ws = await websockets.connect(uri=self.endpoint_uri, loop=self.loop, **self.websocket_kwargs)
+        return self.ws
+
+    async def __aexit__(self, exc_type: Type[BaseException], exc_val: BaseException, exc_tb: TracebackType) -> None:
+        """Close the websocket."""
+        if self.ws is not None and exc_val is not None:
+            try:
+                await self.ws.close()
+            except Exception:
+                pass
```
Comment (michaelhly) [inline solana/rpc/providers/ws.py:64]: Log exceptions on exit. Something like:

```python
except Exception as err:
    log.debug("Exception caught on exit", str(err))
```

You can pass the logger from provider to here.

File: solana/rpc/providers/ws.py [hunk 19]
```diff
@@ -0,0 +1,131 @@
...
+            self.ws = None
+
+
+class Provider(BaseProvider, FriendlyJsonSerde):
```
Comment (michaelhly) [inline solana/rpc/providers/ws.py:68]: Rename to `WSProvider`.

File: solana/rpc/websocket.py [hunk 20]
```diff
@@ -0,0 +1,162 @@
...
+    def account_subscribe(self, pubkey: Union[PublicKey, str]) -> RPCResponse:
```
Comment (michaelhly) [inline solana/rpc/websocket.py:19]: Handle options for `encoding` and `commitment`.

https://docs.solana.com/apps/jsonrpc-api#parameters-44

File: solana/rpc/websocket.py [hunk 21]
```diff
@@ -0,0 +1,162 @@
...
+
+        >>> ws_client = WebSocketClient("ws://localhost:8900") #doctest: +SKIP
+        >>> resp = ws_client.account_subscribe(PublicKey(1)) #doctest: +SKIP
+        >>> account_id = resp['result'] #doctest: +SKIP
+        >>> ws_client.account_unsubscribe(account_id) #doctest: +SKIP
+        {'jsonrpc': '2.0', 'result': True, 'id': 3}
+        """
+        return self._provider.make_request(RPCMethod("accountUnsubscribe"), account_id)
+
+    def program_subscribe(self, pubkey: Union[PublicKey, str]) -> RPCResponse:
```
Comment (michaelhly) [inline solana/rpc/websocket.py:44]: Handle options for `encoding` and `commitment`.

https://docs.solana.com/apps/jsonrpc-api#programsubscribe

File: solana/rpc/websocket.py [hunk 22]
```diff
@@ -0,0 +1,162 @@
...
+
+        >>> ws_client = WebSocketClient("ws://localhost:8900") #doctest: +SKIP
+        >>> resp = ws_client.program_subscribe(PublicKey(1)) #doctest: +SKIP
+        >>> account_id = resp['result'] #doctest: +SKIP
+        >>> ws_client.program_unsubscribe(account_id) #doctest: +SKIP
+        {'jsonrpc': '2.0', 'result': True, 'id': 6}
+        """
+        return self._provider.make_request(RPCMethod("programUnsubscribe"), account_id)
+
+    def signature_subscribe(self, pubkey: Union[PublicKey, str]) -> RPCResponse:
```
Comment (michaelhly) [inline solana/rpc/websocket.py:67]: Handle option for `Commitment`.

File: solana/rpc/websocket.py (hunk 20 above)
Comment (michaelhly) [reply solana/rpc/websocket.py:19]: Example: https://github.com/michaelhly/solana-py/blob/0251648328159533eda1bd92a730f13603315936/solana/rpc/api.py#L63

File: notebooks/Websockets .ipynb [hunk 23]
```diff
@@ -0,0 +1,316 @@
+{
```
Comment (michaelhly) [inline notebooks/Websockets .ipynb:1]: nit: `Websockets .ipynb` -> `Websockets.ipynb`

File: solana/rpc/websocket.py (hunk 15 above)
Comment (menarulalam) [reply solana/rpc/websocket.py]: I believe this particular implementation enters and exits the connection whenever a request is made so this method will always return None.

File: solana/rpc/providers/ws.py [hunk 24]
```diff
@@ -0,0 +1,145 @@
...
+        request_id = next(self._request_counter) + 1
+        self.logger.debug("Making request WebSocket URI: %s, Method: %s, id: %s", self.endpoint_uri, method, request_id)
+        request_data = self.json_encode({"jsonrpc": "2.0", "id": request_id, "method": method, "params": params})
+        if WSProvider._loop:
+            future = asyncio.run_coroutine_threadsafe(
+                self.coro_make_request(cast(bytes, request_data)), WSProvider._loop
+            )
+        return future.result()
+
+    def connection(self) -> Optional[websockets.WebSocketClientProtocol]:
```
Comment (michaelhly) [inline solana/rpc/providers/ws.py:132]: We don't need this, since `conn` is public.

File: solana/rpc/websocket.py [hunk 25]
```diff
@@ -0,0 +1,220 @@
...
+    _data_slice_key = "dataSlice"
+
+    def __init__(self, endpoint: str):
+        """Init API client for websocket."""
+        self._provider = WSProvider(endpoint)
+        self._conn = self._provider.conn.ws
+
+    def connection(self) -> Optional[WebSocketClientProtocol]:
+        """Get the ws connection."""
+        return self._conn
```
Comment (michaelhly) [inline solana/rpc/websocket.py:29]:
```suggestion
        self._provider.conn.ws
```

File: solana/rpc/websocket.py [hunk 26]
```diff
@@ -0,0 +1,220 @@
...
+    """Websocket client class."""
+
+    _comm_key = "commitment"
+    _encoding_key = "encoding"
+    _data_slice_key = "dataSlice"
+
+    def __init__(self, endpoint: str):
+        """Init API client for websocket."""
+        self._provider = WSProvider(endpoint)
+        self._conn = self._provider.conn.ws
```
Comment (michaelhly) [inline solana/rpc/websocket.py:25]:
```suggestion
```

File: solana/rpc/providers/ws.py [hunk 27]
```diff
@@ -0,0 +1,147 @@
...
+        request_id = next(self._request_counter) + 1
+        self.logger.debug("Making request WebSocket URI: %s, Method: %s, id: %s", self.endpoint_uri, method, request_id)
+        request_data = self.json_encode({"jsonrpc": "2.0", "id": request_id, "method": method, "params": params})
+        if WSProvider._loop:
+            future = asyncio.run_coroutine_threadsafe(
+                self.coro_make_request(cast(bytes, request_data)), WSProvider._loop
+            )
+        return future.result()
+
+    def connection(self) -> websockets.WebSocketClientProtocol:
```
Comment (michaelhly) [inline solana/rpc/providers/ws.py:134]:
```suggestion
    def connect(self) -> websockets.WebSocketClientProtocol:
```

To be consistent with https://websockets.readthedocs.io/en/stable/api.html#module-websockets.client

File: solana/rpc/websocket.py [hunk 28]
```diff
@@ -0,0 +1,207 @@
...
+    _encoding_key = "encoding"
+    _data_slice_key = "dataSlice"
+    _filters_key = "filters"
+
+    def __init__(self, endpoint: str):
+        """Init API client for websocket."""
+        self._provider = WSProvider(endpoint)
+        self._conn = self._provider.conn.ws
+
+    def connection(self) -> Optional[WebSocketClientProtocol]:
```
Comment (michaelhly) [inline solana/rpc/websocket.py:27]:
```suggestion
    def connect(self) -> WebSocketClientProtocol:
```

To be consistent with https://websockets.readthedocs.io/en/stable/api.html#module-websockets.client.

**Comment by michaelhly** [comment]
Going to rethink this design. Should not use a separate client to communicate with the WebSocket endpoint. Should design something similar to https://github.com/solana-labs/solana/blob/master/web3.js/src/connection.js#L1364.

**Pull request closed without merging**